frontend/talentsearch/src/js/components/pool/CreateApplication.tsx

```diff
@@ -0,0 +1,141 @@
+   * directly to this, so check if they should
+   * be able to view/apply to it
+   */
+  const isVisible = isAdvertisementVisible(
+    data?.me?.roles?.filter(notEmpty) || [],
+    data?.poolAdvertisement?.advertisementStatus ?? null,
+  );
+  const hasApplied = hasUserApplied(
+    (data?.me?.poolCandidates as Maybe<PoolCandidate>[]) || [],
+    data?.poolAdvertisement?.id,
+  );
+
+  /**
+   * Check to see if the application even exists
+   */
+  const application = data?.me?.poolCandidates?.find((candidate) => {
+    return candidate?.pool.id === data?.poolAdvertisement?.id;
+  });
+
+  /**
+   * Store if the application can be created
+   *
+   * !creating - Not currently running a mutation
+   * !mutationData - The mutation has not previously ran
+   * userId - We need a user ID to run the mutation
+   * id - We need a pool ID to run the mutation
+   * isVisible - Should't run it if user cannot view it
+   * !hasApplied - Users can only apply to a single pool advertisement
+   */
+  const userId = data?.me?.id;
+  const canCreate =
+    !creating && !mutationData && userId && id && isVisible && !hasApplied;
+
+  /**
+   * Actually run the mutation
+   */
+  if (canCreate) {
+    /**
+     * Handle any errors that occur during mutation
+     *
+     * @returns null
+     */
+    const handleError = () => {
+      redirect(poolPath);
+      toast.error(
+        intl.formatMessage({
+          defaultMessage: "Error application creation failed",
+          id: "tlAiJm",
+          description: "Application creation failed",
+        }),
+      );
+      return null;
+    };
+
+    if (userId) {
+      executeMutation({ userId, poolId: id })
+        .then((result) => {
+          if (result.data?.createApplication) {
+            // Redirect user to the application on success
+            redirect(paths.reviewApplication(result.data.createApplication.id));
+            toast.success(
+              intl.formatMessage({
+                defaultMessage: "Application created",
+                id: "U/ji+A",
+                description: "Application created successfully",
+              }),
+            );
+            return null;
+          }
+          return handleError();
+        })
+        .catch(handleError);
+    }
+  }
+
+  // If a pool is not visible, you can create one
+  // So redirect them back to the pool page
+  if (!isVisible) {
+    redirect(poolPath);
+  }
+
+  // If a user has already applied, redirect to the existing application
+  // Or, back to the pool page if we cannot find it
+  if (hasApplied && !creating) {
+    const redirectPath = application
+      ? paths.reviewApplication(application.id)
+      : poolPath;
+    redirect(redirectPath);
+  }
+
+  /**
+   * Render the loading spinner while we do
+   * the necessary work
+   *
+   * Note: This component should always redirect to a path
+   * based on the logic, so no need to render anything but
+   * a loading spinner
+   */
+  return <Loading />;
```

Comment:
So, we can't use any general "else" since that means we would never land on the loading spinner and the page would just flash a bit and could result in some duplicate requests. Though, I have updated the logic a bit to make it a bit cleaner and hopefully reduce the chances of sticking around on this page.

[update logic to prevent staying on loading](https://github.com/GCTC-NTGC/gc-digital-talent/pull/4451/commits/101cddeb2a289909460cbd4b0f96f69c14dd6705)

Reply:
This still gives me the willies.  :worried:   So the first condition is `canCreate` which is great.  Is the second condition `!canCreate`?  That would make it obvious there are no other cases.  If not could you declare a boolean to name what it is?  Then maybe add a third "if" for `if (!canCreate && !secondCondition)` to handle it?

Reply:
Yeah, so I'm not sure what to do here 😅 We can't use any sort of generic redirect fallback because if we do it causes a really bad screen flash, numerous false positive errors and a worse overall UX.

Ideally all of this logic is handled on the backend where we can return an error if all conditions fail, but unfortunately that was not the initial scope of these bugs. I initially only wrote the logic to prevent bugs that already existed with the intention of eventually doing it the "right" way.

**Initial acceptance criteria:**

- Browse opportunities should be available for everyone.
- Apply button redirect user to login page, with post-login-redirect sending you to application process

I'm willing to expand this issue out to move all conditions to the backend and close https://github.com/GCTC-NTGC/gc-digital-talent/issues/4463 at the same time? That way we don't really need to do any logic except make sure the mutation isn't already running in this route.

```tsx
const canCreate = !creating && !mutationData && userId && poolId

if(canCreate) {
  executeMutation({ userId, poolId })
    .then((result) => {
      // Success, move onto application
    })
    .catch((err) => {
      // API throws error so we land here
    });
}

return <Loading />;
```

Reply:
Eh, I think it's one of those things that sounds easy to someone who doesn't fully understand it but it really isn't.  :laughing:   OK, you opened a new ticket for the API guard already.  Maybe add the conditions to the scope of that issue so we can get to the clipped code you have here.  For now I'll approve this one so we can keep moving.

Thanks for indulging me.  :smile:

Reply:
For sure, you brought up a really good point! Honestly, our backlog is getting thin so I may just move onto that issue anyway so we can square this away before direct intake 😜

Reply:
okay, I have updated the issue to hopefully include a summary of this discussion and make it clear.

https://github.com/GCTC-NTGC/gc-digital-talent/issues/4463